numba/array.py

```diff
@@ -0,0 +1,177 @@
+from numba.tracing import trace
+import numpy
+import sys
+
+def _o2s(dtype, shape, order):
+    # convert order parameter to strides
+
+    if dtype is None or shape is None or order is None:
+        return None
+
+    if order == 'F':
+        shape = list(shape)
+        shape.reverse()
+    strides = []
+    itemsize = dtype.itemsize
+    for i in range(len(shape), 0, -1):
+        strides.append(itemsize)
+        itemsize *= shape[i - 1]
+    if order in ('C', None):
+        strides.reverse()
+    return tuple(strides)
+
+def _s2o(dtype, shape, strides):
+    # convert strides parameter to order
+    # Note: strides must correspond to contiguous data layout
+
+    if strides is None or strides[-1] == dtype.itemsize:
+        order = 'C'
+    elif strides[0] == dtype.itemsize:
+        order = 'F'
+    else:
+        raise Error('strides do not correspond to contiguous data layout')
+    s2 = _o2s(dtype, shape, order)
+    if strides != s2:
+        raise Error('strides do not correspond to contiguous data layout')
+    return order
+
+class Array:
+    """An array type that supports host and GPU storage."""
+
+    def __init__(self, obj=None, copy=True,
+                 shape=None, dtype=None, order=None, where='host'):
+        """Construct an Array in the memory space defined by 'where'."""
+
+        assert where in ('host', 'gpu')
+        # we need either a prototype or proper type info
+        assert obj is not None or (shape and dtype)
+        self._host = self._gpu = None
+        self._host_valid = self._gpu_valid = False
+        self._allocate(where, obj, dtype, shape, _o2s(dtype, shape, order), copy)
+        if where == 'host':
+            self._host_valid = True
+            t = self._host
+        else:
+            self._gpu_valid = True
+            t = self._gpu
+        self.shape = t.shape
```

Comment:
You should perhaps make these public attributes properties, so that user code cannot modify them.
(note modifying the `shape` is allowed on Numpy arrays... but it reshapes the array)